core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java

```diff
@@ -76,14 +81,40 @@ public Boolean evaluate(ProjectConfig config, Map<String, ?> attributes) {
         Object userAttributeValue = attributes.get(name);
 
         if (!"custom_attribute".equals(type)) {
-            MatchType.logger.error(String.format("condition type not equal to `custom_attribute` %s", type));
+            logger.warn("Audience condition \"{}\" has an unknown condition type. You may need to upgrade to a newer release of the Optimizely SDK", this.toString());
             return null; // unknown type
         }
         // check user attribute value is equal
         try {
-            return MatchType.getMatchType(match, value).getMatcher().eval(userAttributeValue);
+            Match matchType = MatchType.getMatchType(match, value).getMatcher();
+            Boolean result = matchType.eval(userAttributeValue);
+
+            if (!(matchType instanceof NullMatch) && result == null) {
```

Comment:
I'm opposed to this approach. The `instanceof` and numerous logic branches are hard to follow and logging makes assumptions about the implementation details of matcher. This may be correct right now, but not guaranteed in the future. It smells out of being out of place and not a proper utilization of `MatchType`/`Matcher`/etc

Reply:
Okay, smells a little funky to me too.

The alternative is logging from the matcher itself. Given the type of log message that we want, that means passing down the entire audience condition to the matcher, which I think you'd [been unsure about](https://github.com/optimizely/java-sdk/pull/251#discussion_r244823028). And we'd also want to pass an additional boolean if we want to preserve distinct messages for "no value" and "a null value".

How should we proceed?

Reply:
My opinion (which I will also mention in other SDK's PRs) are that it's okay to pass additional data for logging purposes. @msohailhussain @mikeng13 FYI

Reply:
Sorry, I'm so late to the party.  But, I agree: It stinks.  2 questions on approach:
1.  What's wrong with logging the null messages on the match type and remove the if instanceof?
2.  Why not just create two exceptions and catch for those and log that way?  It is way cleaner than this.